### PR TITLE
moving to a simpler implementation

### PR DESCRIPTION
**Pull request by iimpulse**
- Category implementation requires a graph, returns single category for a term
- method to get the categories that exist generally
- deprecate old classes & tests ( please check I did this right)

### CODE REVIEW

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategory.java [hunk 1]
```diff
@@ -13,60 +13,90 @@
  * gastrointestinal can have the subcategory liver.</p>
  * <p>The main use case for this class is to display a set of Terms that descend from this Category. The class
  * is intended to be used with HpoCategoryMap to display a set of terms that can belong to </p>
+ *
  * @author <a href="mailto:peter.robinson@jax.org">Peter Robinson</a>
  * @version 0.1.14
  */
+@Deprecated(since = "3.0.0", forRemoval = true)
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategory.java:20]: The class should be deprecated as of this version (`2.0.6`). I'll push a change where I'll update this, and I'll also add @deprecated JavaDoc tag where we will suggest the replacement. The suggestion makes it easier for the user to update the code as soon as he notices the deprecation.

Otherwise, I believe the remaining changes of this class are just formatting updates. Please let me know if I'm wrong.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class HpoCategoryLookup {
+
+  Logger LOGGER = LoggerFactory.getLogger(HpoCategoryLookup.class);
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java:16]: Logger should probably be `private static final`.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  public Term[] categories(){
+    return categories;
+  }
+
+  /**
+   * This finds the "best" category for a term. The categories do not cover all terms.
+   * @param termId
+   * @return the Term for a category
+   */
+  public Term getPrioritizedCategory(TermId termId) {
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java:69]: Consider changing the return type to `Optional<TermId>` unless you hate optionals. :) I kind of like them since they communicate that the result is nullable.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    if(categories.isEmpty()){
+      LOGGER.warn("Could not find category for {}", termId);
+      return null;
+    } else if (categories.stream().anyMatch(t -> t.equals(NEOPLASM_ID))) {
+      return NEOPLASM_ID;
+    }
+    return categories.get(0);
+  }
+
+  private List<Term> getCategoriesForId(TermId termId){
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java:80]: The usage of this method suggests you only get the first term. Consider not collecting into the full list if all you need is the first element.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import org.monarchinitiative.phenol.ontology.data.TermId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class HpoCategoryLookup {
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java:14]: Two points wrt. this class.

I think it would be better to make this class to just work with the `TermId`s. Then less information is hard-coded (no labels). This assumes that the user of this class has `MinimalOntology` around, and can translate the `TermId` into its label (EN, CZ, NE, whatever language). In result `getPrioritizedCategory` method will return `List<TermId>` instead of `Term`s.

Second, I think it would be better to ask for the categories via the constructor and not hard-code them as a field. This will help us change the categories in the future, and make the class more usable, in general. Asking for a `List<TermId>` is probably OK, I am not sure having `TermId[]` has any advantages. In result, the `HpoCategoryLookup` will have two private final fields: `OntologyGraph<TermId> graph` and `List<TermId> categories`.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryMap.java [hunk 6]
```diff
@@ -24,6 +24,7 @@
  * @author <a href="mailto:peter.robinson@jax.org">Peter Robinson</a>
  * @version 0.1.7
  */
+@Deprecated(since = "3.0.0", forRemoval = true)
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryMap.java:27]: I'll update this too.

File: phenol-annotations/src/test/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookupTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import org.monarchinitiative.phenol.ontology.data.Term;
+import org.monarchinitiative.phenol.ontology.data.TermId;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class HpoCategoryLookupTest {
```
Comment (ielis) [inline phenol-annotations/src/test/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookupTest.java:15]: We should have `Arachnodactyly` in `hpo_toy.json`. Can you pls add a test that checks that we assign a correct category to `Arachnodactyly`?

File: phenol-annotations/src/test/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryMapTest.java [hunk 8]
```diff
@@ -15,6 +15,7 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
+@Deprecated
```
Comment (ielis) [inline phenol-annotations/src/test/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryMapTest.java:18]: I think we don't have to deprecate tests. Removing the tested class will result in compile error which will suggest us to trash it.

File: phenol-annotations/src/test/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryMapTest.java [hunk 9]
```diff
@@ -136,6 +137,24 @@ public void testTwoCategories() {
     assertTrue(catlist.stream().anyMatch(cat -> cat.id().equals(INHERITANCE_ID)));
   }
 
+  /**
+   * Add terms from INHERITANCE and from LIMBS_ID and CATEGORY match
+   */
+  @Test
+  public void testCategoryExact(){
```
Comment (ielis) [inline phenol-annotations/src/test/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryMapTest.java:144]: This test seems to be new. Does it accomplish anything? Why did you add a test that will eventually be removed? A mis-commit?

**Review by ielis — CHANGES_REQUESTED** [review]
Hi @iimpulse
I'd like to ask you to consider my suggestions mainly on the `HpoCategoryLookup`. 

I added few notes regarding deprecation. I'll push a commit with few more lines of code.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategories.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
+package org.monarchinitiative.phenol.annotations.formats.hpo.category;
+
+import org.monarchinitiative.phenol.ontology.data.Term;
+import org.monarchinitiative.phenol.ontology.data.TermId;
+
+public class HpoCategories {
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategories.java:6]: Pls add a private no-op constructor to emphasize this is a utility class.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class HpoCategoryLookup {
+
+  private final Logger LOGGER = LoggerFactory.getLogger(HpoCategoryLookup.class);
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java:17]: Logger should be `static`.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+  /**
+   * This finds the "best" category for a term. The categories do not cover all terms.
+   * @param termId
+   * @return the Term for a category
+   */
+  public Optional<Term> getPrioritizedCategory(TermId termId) {
+    List<Term> categories = getCategoriesForId(termId);
+    if(categories.isEmpty()){
+      LOGGER.warn("Could not find category for {}", termId);
```
Comment (ielis) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/formats/hpo/category/HpoCategoryLookup.java:36]: Logging a missing category is probably not necessary here.

**Review by ielis — APPROVED** [review]
Hi @iimpulse 
looks good to me, thanks a lot!
